internal/tofu/context_import.go

```diff
@@ -71,80 +72,87 @@ func (i *ImportTarget) IsFromImportCommandLine() bool {
 // Import targets' addresses are not fully known from the get-go, and could only be resolved later when walking
 // the graph. This struct helps keep track of the resolved imports, mostly for validation that all imports
-// have been addressed and point to an actual configuration
+// have been addressed and point to an actual configuration.
+// The key of the map is a string representation of the address, and the value is an EvaluatedConfigImportTarget.
 type ImportResolver struct {
 	mu      sync.RWMutex
-	imports map[ResolvedConfigImportsKey]EvaluatedConfigImportTarget
+	imports map[string]EvaluatedConfigImportTarget
 }
 
 func NewImportResolver() *ImportResolver {
-	return &ImportResolver{imports: make(map[ResolvedConfigImportsKey]EvaluatedConfigImportTarget)}
+	return &ImportResolver{imports: make(map[string]EvaluatedConfigImportTarget)}
 }
 
 // ResolveImport resolves the ID and address (soon, when it will be necessary) of an ImportTarget originating
 // from an import block, when we have the context necessary to resolve them. The resolved import target would be an
 // EvaluatedConfigImportTarget.
 // This function mutates the EvalContext's ImportResolver, adding the resolved import target
 // The function errors if we failed to evaluate the ID or the address (soon)
-func (ri *ImportResolver) ResolveImport(importTarget *ImportTarget, ctx EvalContext) error {
-	importId, evalDiags := evaluateImportIdExpression(importTarget.Config.ID, ctx)
-	if evalDiags.HasErrors() {
-		return evalDiags.Err()
+func (ri *ImportResolver) ResolveImport(importTarget *ImportTarget, ctx EvalContext) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
+	// The import block expressions are declared within the root module.
+	// We need to explicitly use the context with the path of the root module, so that all references will be
+	// relative to the root module
+	rootCtx := ctx.WithPath(addrs.RootModuleInstance)
+
+	importId, evalDiags := evaluateImportIdExpression(importTarget.Config.ID, rootCtx)
+	diags = diags.Append(evalDiags)
+	if diags.HasErrors() {
+		return diags
 	}
 
-	// TODO - Change once an import target's address is more dynamic
-	importAddress := importTarget.Config.To
+	importAddress, addressDiags := rootCtx.EvaluateImportAddress(importTarget.Config.To)
+	diags = diags.Append(addressDiags)
+	if diags.HasErrors() {
+		return diags
+	}
 
 	ri.mu.Lock()
 	defer ri.mu.Unlock()
 
-	resolvedImportKey := ResolvedConfigImportsKey{
-		AddrStr: importAddress.String(),
-		ID:      importId,
+	resolvedImportKey := importAddress.String()
+
+	if importTarget, exists := ri.imports[resolvedImportKey]; exists {
+		return diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  fmt.Sprintf("Duplicate import configuration for %q", importAddress),
+			Detail:   fmt.Sprintf("An import block for the resource %q was already declared at %s. A resource can have only one import block.", importAddress, importTarget.Config.DeclRange),
```

Comment:
Just for my curiousity, can someone declare an import block AND an import through the CLI somehow?

I wonder if we need to handle it with a more generic error message

Reply:
It cannot be done. When you run an import using the CLI, all other `import` blocks in the configuration are ignored.